src/CodeGen_LLVM.cpp

```diff
@@ -2807,6 +2808,147 @@ void CodeGen_LLVM::visit(const Call *op) {
                 }
             }
         }
+    } else if (op->is_intrinsic(Call::declare_struct_type)) {
+        // Declares a struct type. Returns a null pointer of the new type.
+        internal_assert(op->args.size() >= 2);
+
+        const StringImm *str_imm = op->args[0].as<StringImm>();
+        internal_assert(str_imm != nullptr);
+        std::string name = str_imm->value;
+
+        const int64_t *mode = as_const_int(op->args[1]);
+        internal_assert(mode != nullptr);
+
+        llvm::Type *struct_type;
+        if (*mode == 1) {
+            vector<llvm::Type *> types(op->args.size() - 2);
+            for (size_t i = 2; i < op->args.size(); i++) {
+                types[i - 2] = codegen(op->args[i])->getType();
+            }
+            struct_type = (llvm::Type *)llvm::StructType::create(*context, types, "struct." + name);
+        } else {
+            struct_type = get_llvm_struct_type_by_name(module.get(), ("struct." + name).c_str());
+            internal_assert(struct_type != nullptr) << "Failed to find structure " << name << "\n";
+        }
+        value = llvm::ConstantPointerNull::get(struct_type->getPointerTo());
```

Comment:
What are we going to do here if llvm finally removes types from pointers? Have a separate symbol table for types perhaps?

Reply:
a) The main reason I'm doing this is to get LLVM IR to type check. I thought it would be safer, but we might just cast to the expected type at the point of use.

b) We can just track the type ourselves. Either by using our own wrapper on llvm::Value and an llvm::Type to carry it around, or via an adjunct table.

c) Do all the work to make sure the full C++ typed pointer is carried in the Halide IR.

It is all internal to our IR and changing it shouldn't cause user visible changes.